docs/source/tutorial/codec.rst

```diff
@@ -0,0 +1,365 @@
+
+Plugins: make your own
+----------------------
+
+It's simple to make your own additional plugins. They can be defined in several places: job configs, macros,
+and any file on the classpath, which you would use the "include" keyword to import. When a plugin becomes general
+purpose, it is often moved to extra-filters or to the hydra config files directly.
+
+A good place to start learning how to make plugins is looking at the
+`extra-filters <https://github.com/addthis/hydra/blob/master/hydra-filters/src/main/resources/extra-filters-src.conf>`_
+in the hydra git repository. Here's the basic structure of how to define a plugin::
+
+   # example bundle filter
+   # ("bundle-filter" is the type of pluggable thing we are creating)
+   plugins.bundle-filter {
+      # plugin name
+      field-to-field2 {
+         _class: chain # name of base class. chain, and field are common use cases
+         # any special properties like _primary
+         # properties of the base class you want to override in the plugin
+         # let's copy the value in FIELD to FIELD2,
+         # but only if we find "foobar" as a substring in the field.
+         filter: [
+            {from:"FIELD", to:"FIELD2", require.contains:"foobar"}
+         ]
+      }
+   }
+
+Plugins can also be defined by referencing a new class directly, for example creating MyCoolCoolBundleFilter java class
+and then setting it up in a conf file like extra-filters, or in the hydra conf files directly.
+
+hocon "macros" which are just functions (global vs underscores)
+
+Substitution: What can underscores do for you?
+----------------------------------------------
+
+In addition to the plugin system which allows you to create new plugins for pluggable types, HOCON job configs also
+allow you to do `substitution <https://github.com/typesafehub/config#uses-of-substitutions>`_. There is one
+important thing to know about substitution in hydra: you need to put things you plan to use for substitution
+in either the global namespace (`global.thing-to-substitute`) or prefix it with an underscore (`_thing-to-substitute`).
+
+The reason this matters is that normally in a hydra job, the parser tries to read everything into a java object. The
+only way it knows that something is not supposed to be coerced into a java object is for it not to be available to
+parse. The global namespace removes it from the parsing by placing it elsewhere in the tree. And any underscore-prefixed
+property is ignored for java objects, as it is used for special properties like _class, _primary, etc.
+
+With HOCON substitution, you basically get job config reuse without duplicating your code. Here is an example
+of this in practice, lifted from a real hydra job::
+
+   _uidcount {
+       uid.count:{key:"UID", max:50000000}
+       uidage.count:{key:"UID_GOOD", max:50000000}
+       uidage48.count:{key:"UID_SUPER_GOOD", max:50000000}
+   }
+
+   output.tree.root: [
+      {const:"ROOT"}
+      {branch:[
+         [
```

Comment:
In other words, The only reason you need nested arrays is if you want each sibling to be more than one node deep. Got it.